[BT-186] Martha returns signed URLs for open-access BDC data

kshakir opened the pull request.

mcovarr committed:
Clarifying instructions.

breilly2 committed:
WIP

mcovarr committed:
wip in case I need to disappear again

breilly2 committed:
Test runs and fails

breilly2 committed:
added temp test script thingy

breilly2 committed:
WIP making test green

mcovarr committed:
checkpoint

mcovarr committed:
on second thought...

kshakir committed:
WIP wired in the fourth call to BDC

breilly2 committed:
It's green!!!

kshakir committed:
WIP debugging more tests

breilly2 committed:
WIP

mcovarr committed:
wip

kshakir committed:
wip down to 15 failing unit tests!

kshakir committed:
WIP more passing tests

breilly2 committed:
fixed a test

mcovarr committed:
fix another

mcovarr committed:
fix another

mcovarr committed:
one more

mcovarr committed:
just like the last one

mcovarr committed:
one more...

breilly2 committed:
fixed a test

breilly2 committed:
fixed a test

breilly2 committed:
fixed two tests

breilly2 committed:
WIP: test is still broken

mcovarr committed:
one more (not the WIP one, that's still broken)

mcovarr committed:
Is minnie-kenny ok with this?

mcovarr committed:
We do expect a bearer token since HCA data was moved from UCSC which did not require one to Jade which does.

mcovarr committed:
add a test to get some missing unhappy path test coverage

mcovarr committed:
another negative test, fix the previous negative test

mcovarr committed:
cleanup

breilly2 committed:
WIP

mcovarr committed:
test fixes

mcovarr committed:
linter fixes

breilly2 committed:
fixes after rebase

kshakir committed:
Multiple updates:
- JDR hosts now specify that they do NOT want to use an access_id
- Removed quotes from example DRS response field names
- Renamed HCA variable names to match the change in protocol from DOS to DRS
- Removed handling for retired HCA DOS server
- Removed unused ALL_FIELD_DOS
- All martha_v3 tests prefixed with `martha_v3` for better test filtering for `ava -m 'martha_v3*'`
- All martha_v3 tests consistently use test.serial()

kshakir committed:
Multiple updates:
- Added trailing commas to help future updates
- Updated accessUrl.headers JSDoc
- Renamed httpsUrlGenerator to generateMetadataUrl
- Updated GCF deployment to nodejs 12
- Switched url.format to use new WHATWG URL
- Added tests for hypothetical case of a DOS/DRS URI with a query string that also needed a generated access url

kshakir committed:
Merge branch 'dev' into bt_170_access_url

kshakir commented on martha/martha_v3.js:
```diff
@@ -398,31 +424,66 @@ async function marthaV3Handler(req, res) {
         return;
     }
 
+    // do more stuff here, i.e. get signed URL
+    // first, fetch access token from bond
+    // TODO: allow this to be not done based on requested data
+    let accessToken;
+    if (bondAccessTokenUrl && accessId) {
```
- [x] Replace the comments with something more descriptive
- [x] Don't access this `if` if the client hasn't requested the `accessUrl`

kshakir commented on martha/martha_v3.js:
```diff
@@ -398,31 +424,66 @@ async function marthaV3Handler(req, res) {
+
     let bondSA;
-    if (bondUrl && overlapFields(requestedFields, BOND_FIELDS)) {
+    if (bondSAKeyUrl && overlapFields(requestedFields, BOND_FIELDS)) {
```
`BOND_FIELDS` should be renamed. It is now the fields that imply one should contact bond for the SA key

kshakir committed:
Multiple updates:
- Removed TODO
- Changed some double quotes to single quotes
- Removed mob scripts from package.json
- Added `accessUrl` to README.md
- Updated docs regarding default fields
- Updated docs regarding node 12

mcovarr commented on martha/martha_v3.js:
```diff
@@ -76,11 +80,24 @@ const DG_COMPACT_KIDS_FIRST = 'dg.f82a1a';
+    }
+}
+
+/* Returns undefined if the matching access method does not have an access_id. */
```
... or if the `accessMethodType` is falsy or the `drsResponse` is falsy

kshakir committed:
Multiple updates:
- Return the fileName from the metadata first, then the accessUrl second, and lastly the gsUrl/gsUri
- Updated README.md to reflect that the fileName may not come from the gsUri

kshakir committed:
During martha_v3 testing, don't globally return the googleSAKeyObject on the second call to getJsonFrom()

mcovarr commented on README.md:
```diff
@@ -67,6 +68,12 @@ Example response for /martha_v3:
+      "headers": {
+        "Authorization": "Basic Z2E0Z2g6ZHJz"
+      }
+    },
```
nice!

mcovarr commented on common/helpers.js:
```diff
@@ -266,14 +268,27 @@ const promiseHandler = (fn) => (req, res) => {
  * Extracts the bucket and path from a Google Cloud Storage URL
  *
- * @param uri The GCS url
+ * @param {?string} uri The GCS url
```
uri and url being used interchangeably

kshakir replied:
_Matches the code._ /s 🤦

Mini rant: GCS technically [uses a URI](https://cloud.google.com/bigquery-transfer/docs/cloud-storage-transfer-overview#google-cloud-storage-uri) while consistently a lot of code around the various DSP repos "wings it" by using classes like the [WHATWG URL](https://nodejs.org/api/url.html#url_the_whatwg_url_api) to parse the values.

And then of course these are not `URL` objects but `string`...

kshakir committed:
Multiple updates:
- Dodged the eslint complexity for martha_v3 by moving the response generation into a function
- When the accessUrl is requested, query the expected endpoints (now tested too)
- Renamed martha_v3 field constants
- Removed more quotes from object field names
- Stopped stubbing some calls that weren't needed
- Converted some strings to/from double quotes
- Added more trailing commas to reduce future diffs

kshakir committed:
Multiple updates:
- Added todo regarding method complexity
- Retrieve drs signed url if the fileName is requested and the DRS metadata doesn't have fields to generate that value
- More quotes converted to single quotes

kshakir committed:
Do not overwrite the name field returned from DRS metadata

kshakir committed:
Fixed warning regarding regex

kshakir commented on martha/martha_v3.js:
```diff
@@ -406,23 +486,67 @@ async function marthaV3Handler(req, res) {
+    // Try to retrieve the file name from the initial DRS response
+    let fileName = getDrsFileName(drsResponse);
+
+    // If we do NOT have the fileName yet, we will try to get that field from the accessUrl
+    const accessIdFields = fileName ? MARTHA_V3_ACCESS_ID_FIELDS : MARTHA_V3_ACCESS_ID_FN_FIELDS;
```
To discuss: It should be possible to get the `fileName` only from the metadata without requesting the signed URL.

kshakir replied:
Decision: Return `null` and do not access the signed URL to sleuth out the file name

kshakir commented on martha/martha_v3.js:
```diff
@@ -174,7 +232,7 @@ function getHttpsUrlParts(url) {
     // The many different ways a DOS/DRS may be "compact", in the order that the should be tried
     const cibRegExps = [
         // Non-W3C CIB DOS/DRS URIs, where the `dg.abcd` appears more than once
-        /(?:dos|drs):\/\/(?<host>dg\.[0-9a-z-]+)(?<separator>:)(?:\k<host>)\/(?<suffix>[^?]*)(?<query>\?(.*))?/i,
+        /(?:dos|drs):\/\/(?<host>dg\.[0-9a-z-]+)(?<separator>:)\k<host>\/(?<suffix>[^?]*)(?<query>\?(.*))?/i,
```
FYI IntelliJ suggested this edit.

breilly2 replied:
Makes sense. It would throw off some match groups if referenced by position, but we're naming them all so it should be okay.

kshakir commented on deploy.sh:
```diff
@@ -108,13 +108,13 @@ docker run \
+      gcloud beta functions deploy fileSummaryV1 --trigger-http --source=. --runtime nodejs12 \\
         --allow-unauthenticated --project ${DEPLOY_PROJECT_NAME} &&
-      gcloud beta functions deploy getSignedUrlV1 --trigger-http --source=. --runtime nodejs10 \\
+      gcloud beta functions deploy getSignedUrlV1 --trigger-http --source=. --runtime nodejs12 \\
```
FYI: This aligns "real" GCF with the nodejs version used by the emulator run inside FiaB containers: https://github.com/broadinstitute/martha/blob/3.5.0/docker/Dockerfile#L2

kshakir committed:
Updated comments

kshakir committed:
Updated logging and variable scope.

kshakir committed:
Comment updates.

mcovarr commented on test/martha/martha_v3.test.js:
```diff
@@ -63,36 +69,49 @@ const mockResponse = () => {
+    return { url: `https://storage.googleapis.com/${gsUrl.hostname}${gsUrl.pathname}?sig=ABC` };
+}
+
 const bondRegEx = /^https:\/\/([^/]+)\/api\/link\/v1\/([a-z-]+)\/serviceaccount\/key$/;
```
plz bondServiceAccountRegEx

mcovarr commented on test/martha/martha_v3.test.js:
```diff
@@ -526,22 +780,27 @@ test.serial('martha_v3 parses BDC staging response correctly', async (t) => {
-    t.falsy(getJsonFromApiStub.firstCall.args[1]); // no auth passed
-    const requestedBondUrl = getJsonFromApiStub.secondCall.args[0];
+    t.falsy(getJsonFromApiStub.getCall(0).args[1]); // no auth passed
+    const requestedBondUrl = getJsonFromApiStub.getCall(3).args[0];
```
hey what happened to calls 1 and 2

kshakir commented on test/martha/martha_v3.test.js:
```diff
@@ -155,53 +176,244 @@ test.serial('martha_v3 resolves successfully and ignores extra data submitted be
+
+    t.is(
+        getJsonFromApiStub.getCall(2).args[0],
+        'https://staging.gen3.biodatacatalyst.nhlbi.nih.gov/ga4gh/drs/v1/objects' +
```
Replace these hostnames with constants as is used in the other matches.

kshakir committed:
Refactor martha_v3 handler into smaller functions (#211)

kshakir committed:
Renamed bond variables with regards to SA Key vs Access Token

kshakir committed:
Use constants for BDC

kshakir committed:
Added checks for access token and access URL

kshakir committed:
Return null for fileName when only an access id is returned in the metadata

kshakir committed:
Variable name typo

kshakir committed:
Don't pass variables just for delayed logging

mcovarr commented on test/martha/martha_v3.test.js:
```diff
@@ -354,25 +351,8 @@ test.serial('martha_v3 calls the correct endpoints when only the fileName is req
-    );
-    t.is(getJsonFromApiStub.getCall(2).args[1], `Bearer ${bondAccessTokenResponse.token}`);
+    sinon.assert.callCount(getJsonFromApiStub, 1); // Bond was not called to retrieve the googleServiceAccount
+    t.deepEqual({ ...result }, { fileName: null });
```
TOL is `{ ...result }` not the same as `result`?

kshakir replied:
In this case, probably. There were (are?) similar test cases where `deepEqual` tested the [equality of the wrapper](https://github.com/avajs/ava/blob/main/docs/03-assertions.md#deepequalvalue-expected-message).

For consistency and because a lot of these lines are later copy-pasted I used the same "erase the wrapper type" syntax here.

breilly2 commented on test/martha/martha_v3.test.js:
```diff
@@ -107,99 +135,272 @@ test.serial('martha_v3 resolves a valid DOS-style url', async (t) => {
+});
+
+test.serial('martha_v3 calls the correct endpoints when only the fileName is requested and the metadata contains a gs url', async (t) => {
+    const drsResponse = bdcDrsResponseCustom({
```
This results in an unmodified `bdcDrsResponse`. As a result, I think this behavior is already covered by `martha_v3 parses BDC response correctly`.

kshakir replied:
Yes, true today, and happy to remove this as a separate test for now. If/when BDC changes what fields get populated we can change both `bdcDrsResponse` and `bdcDrsResponseCustom` at that time.

breilly2 commented on test/martha/martha_v3.test.js:
```diff
@@ -351,321 +552,449 @@ test.serial('martha_v3 should return the underlying status if Data Object resolu
+    getJsonFromApiStub.onCall(3).resolves(googleSAKeyObject);
     const response = mockResponse();
     await marthaV3(
         mockRequest({ body: { 'url': `dos://${config.HOST_CRDC_STAGING}/206dfaa6-bcf1-4bc9-b2d0-77179f0f48fc` } }),
```
Should this be `drs://`?

kshakir replied:
Despite attempts to remove DOS from the internet, it did exist and still does.

kshakir committed:
Removed overlapping test of BDC default fileName resolution

breilly2 reviewed (approved):
👍

kshakir committed:
Only return `accessUrl` for BDC

kshakir committed:
Merge branch 'dev' into bt_170_access_url
